### DISCUSSION TITLE
[@types/nightwatch] - Missing Support for Async/Await Element Commands

### DISCUSSION THREAD

**Discussion by KeionneDerousselle** [DefinitelyTyped/DefinitelyTyped#61801] [Issues with a @types package] [answered]
There are no type definitions to support async/await element commands such as `getText`. There are only definitions for the callback versions of element commands. 

For example, the current type definitions define `getText` as:
```ts
    getText(
        selector: Definition,
        callback?: (this: NightwatchAPI, result: NightwatchCallbackResult<string>) => void,
    ): this;

    getText(
        using: LocateStrategy,
        selector: Definition,
        callback?: (this: NightwatchAPI, result: NightwatchCallbackResult<string>) => void,
    ): this;
```

But the comments right above this type definition state that you should be able to `await` `getText` and have it return a `string`. This is also what the [Nightwatch documentation](https://nightwatchjs.org/api/getText.html) and [their tests](https://github.com/nightwatchjs/nightwatch/blob/da1bdc9c3d5297468828a88db8f6c6bbb66aa305/test/sampletests/es6await/webdriver/getText/getTextES6.js#L21) suggests.

```ts
  demoTestAsync: async function(browser) {
    const result = await browser.getText('#main ul li a.first');
    console.log('getText result', result);
  }
```

Is there any reason we can't change the types for element commands to match the following:
```ts
    getText(
        selector: Definition,
        callback: (this: NightwatchAPI, result: NightwatchCallbackResult<string>) => void,
    ): this;

    getText(
        using: LocateStrategy,
        selector: Definition,
        callback: (this: NightwatchAPI, result: NightwatchCallbackResult<string>) => void,
    ): this;

    getText(
      selector: Definition,
    ): Promise<string>;
```

[a comment by an automated account is not shown]

**Comment by lloiser** [comment]
For `getText` this probably will work because it only makes sense to use the `callback` to get the text or the value returned by the promise it returns in async tests.

However, I think this will not work for other commands that don't really have a return value like `click` or `waitForElementVisible`.
Based on your example the `click` command would then look something like this:
```ts
interface Commands {
  click(selector: Definition, callback?: (this: NightwatchAPI, result: NightwatchCallbackResult<null>) => void): this;
  click(selector: Definition): Promise<void>;
}
```
As far as I know, it is impossible to distinguish these two definitions if you do not provide the callback argument (which will be 99% of the `click` usages). Check out this simplified example:
https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIHsC2m4gCYDOyA3gFDIXIIA2wCA1gBQETUQJjpQBcyBYUUAHMANFTjVqAI0T0A-L0ZQIBAK7UwvEOuoBKZAF4AfMgBu6YHl28wAC2AEA3OUo06TFmw5de-QSCFrZAAFKCwHCAAec0sjZwBfUlIYVRAOYHQQZFUWVAlpWUYEXgxsXEJ9MkoqADo3BkYAIkaxJRV1MH1jEhdqigcAOR02tQ1dZ2r48dJEuAIATzTkFLSwDKy50PCWIpKsHHwCSt6qTP5kZVGwQ2Q4AHc4YGuEOtoG5unqhy3MCJGO6aJZKpdKZZDfMK-HamXg-CKRXDzIyVRIrUFZQbDGHIbSSFGkIA

The next problem is that the promise still should support chaining as far as I understand it. So simply returning the value of the command (`string` in case of `getText`) is not sufficient.

I have already mentioned this here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60170#discussion_r869148364 but it seems that my suggested type did not work and needs more to get this working nicely.

**Reply by KeionneDerousselle** [reply]
Is it possible to just remove the `?` from the callback definition. In other words, provide two function definitions for `click` and `getText`. The API does not always return `this`. It returns the value or void when the function is asynchronous. It returns `this` for chaining when the function has a callback. 

```ts
interface Commands {
  click(selector: Definition, callback: (this: NightwatchAPI, result: NightwatchCallbackResult<null>) => void): this;
  click(selector: Definition): Promise<void>;
}
```
I'm pretty sure, might be wrong about this, that the library always return `this` for the callback version and `Promise<T>` for the async version. 

Whether or not the library _should_ return these inconsistent return values is another story...Ideally it would return the same result, whether using the callback or not.

**Reply by lloiser** [reply]
> Is it possible to just remove the ? from the callback definition

I would not do this as this probably will cause a lot of typescript errors in existing tests.

> The API does not always return this. It returns the value or void when the function is asynchronous. It returns this for chaining when the function has a callback.

Both the `callback` and promise code support chaining: https://nightwatchjs.org/guide/writing-tests/using-es-6-async-await.html#overview
> Starting Nightwatch version 1.7, chaining the API commands when using an async function is also supported.

So both of these should work (not tested though, I have only used the callback code so far):
```ts
browser
  .waitForElementVisible("button")
  .click("button")
  .waitForElementVisible("p")
  .getText("p", (result) => {
    console.log(result.value)
  })
```
```ts
const result = await browser
  .waitForElementVisible("button")
  .click("button")
  .waitForElementVisible("p")
  .getText("p")
console.log(result)
```
And I think this is the tricky part in making these types work as expected.
Maybe there is no "nice" solution to keep both callback and promise commands together in the same type definitions.
Maybe 2 different types for the browser, one for the callback based code (`NightwatchBrowser`) and one for the promise based code (e.g. `NightwatchBrowserPromise`) are needed in order to keep the API type definitions "clean" 🤷

**Comment by garg3133** [comment, marked as answer]
As @lloiser mentioned, it might make sense to not have `?` in the callback function in case of commands like `getText` because if someone is using `getText` command, they'd want to do something with the text returned, and so they'd always pass a callback when writing normal test (without async/await). So, in case of `getText`, we can have the type definition as:

```ts
interface Commands {
  getText(selector: Definition, callback: (this: NightwatchAPI, result: NightwatchCallbackResult<string>) => void): this;
  getText(selector: Definition): Promise<string>;
}
```

But in case of other commands like `click`, `waitForElementVisible`, `navigateTo`, etc., we can't enforce users to always pass a callback when writing the tests normally because these commands don't actually return anything to have a callback for. For instance, while writing normal tests, you'd almost always write:
```js
it('demo test', function() {
  browser
    .navigate('https://nightwatchjs.org')
    .waitForElementVisible('body')
    .someOtherCommands();
});
```
and not
```js
it('demo test', function() {
  browser
    .navigate('https://nightwatchjs.org', (result) => {
      // result.value is null
    })
    .waitForElementVisible('body' (result) => {
      // result.value is null
    })
    .someOtherCommands();
});
```

So, if we make passing callback mandatory in normal cases also, either user would need to pass empty callbacks (`() => {}`) to every command like `click` or they won't be able to chain the commands.

And then there's also a catch with commands like `getText` due to which we cannot use the type definitions suggested by OP in these commands as well. Users can actually pass a callback to commands when writing tests in async/await fashion as well and if the callback returns a promise, the result of that promise will be resolved by the command.

```js
it('demo test', async function() {
  const result = await browser.getText('#main ul li a.first', function(result) {
    const textLength = result.value.length;

    return Promise.resolve(textLength);
  });

  console.log(result);  // will return the length of the text
  console.log(typeof result);  // 'number'
});
```
Doc: https://nightwatchjs.org/guide/writing-tests/using-es-6-async-await.html#using-callbacks-with-async

---

> I'm pretty sure, might be wrong about this, that the library always return this for the callback version and Promise<T> for the async version.

This is not entirely true. The commands always return `this` no matter whether the user is using async/await or not (you can also chain the commands when using async/await). Whether the command will return a promise or not depends on whether the parent function (function passed as the second argument to `it`) is `async` or not.

If the parent function is async, all the commands will always return a promise (along with `this`), no matter whether the user is using a callback or not. And if the parent function is not async, the commands will only return `this`.

There's also a global property called `alwaysAsync` which can be set to `true` after which the commands will always return a promise along with `this` no matter whether the outer function is `async` or not.

---

Now coming to the answer to the main discussion, we are right now working on adding the support for async/await in all the commands. The method we are using to solve this is in fact the one suggested by @lloiser [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60170#discussion_r892417555).

The only downside with this solution is that Typescript will always show the commands to be returning a promise no matter whether the user is using async/await or not. But this is not a big issue as users won't be able to `await` the commands without putting an `async` on the outer function (doing which will make the code valid).

The only single problem with this solution is that TypeScript will now allow users to do something like this:

```js
it('demo test', function() {
  browser.navigateTo('https://nightwatchjs.org').then((result) => {
    // something
  });
});
```
which is invalid in JavaScript as the `navigateTo` command won't be returning a promise at this time. But since this has always been invalid in Nightwatch and it will be invalid when users run the above test, this doesn't seem a big issue nor a breaking change.

---

Now coming to the progress on the async/await issue, I am actively working on this. Right now, I'm fixing the types for the CDP commands along with adding async/await support to them. I thought if I'm going through all the commands to add the async/await support, why not check and fix the types also for all the commands along the way? But now it seems to me that adding the support for async/await first makes much more sense and refactoring of types can wait for later.

**Reply by KeionneDerousselle** [reply]
@garg3133 That was my suggestion to make the callback required and not @lloiser 's.

**Reply by garg3133** [reply]
@KeionneDerousselle But I'm saying the opposite, we cannot make the callback mandatory as explained in the second paragraph.

> But in case of other commands like `click`, `waitForElementVisible`, `navigateTo`, etc., we can't enforce users to always pass a callback when writing the tests normally because these commands don't actually return anything to have a callback for.

**Reply by KeionneDerousselle** [reply]
Also, @garg3133, I'm not sure what you're saying about functions like `navigate`. Those functions don't have a callback parameter. The aren't asynchronous either. My suggestion is that:

1) any asynchronous function that should return a value, should return `Promise<T>`.  For example, all commands that begin with `get`.
 
2) Any asynchronous functions that don't return a value, but instead are chainable should return `this`. For example, `waitForElementVisible` should return `Promise<NightwatchBrowser>` or "this".

3) Any synchronous functions that have a callback, should return `this`. They should provide the `NightwatchCallbackResult<T>` as the result param in the callback function.

I get that these types don't make actual sense, but they should just follow what the actual library is returning. They even mention this when talking about how to create your own custom commands. 

Please take a look at some of their unit tests which assert that the return value for asynchronous `get` functions return a value and not the Nightwatch instance. 

https://github.com/nightwatchjs/nightwatch/blob/da1bdc9c3d5297468828a88db8f6c6bbb66aa305/test/sampletests/es6await/webdriver/getText/getTextES6.js#L21

https://github.com/nightwatchjs/nightwatch/blob/2d30d311f444fcc7721d64466f794557c62c7ea8/test/apidemos/elements/elementGlobalTest.js#L49

**Reply by KeionneDerousselle** [reply]
@garg3133 Sorry, I just re-read your explanation. This makes more sense.

**Reply by garg3133** [reply]
> Those functions don't have a callback parameter. The aren't asynchronous either.

All the commands in Nightwatch have a callback parameter and all commands are asynchronous in nature (Nightwatch allows users to write the test code in a synchronous fashion but all the commands are actually executed asynchronously in Nightwatch).

Here's the documentation of `navigateTo` command which shows the callback parameter: https://nightwatchjs.org/api/navigateTo.html

Also, all the commands in Nightwatch are chainable (no matter whether we are using them synchronously or asynchronously). So, we always need to return `this` in all cases. The only thing to decide is whether to return a promise or not, which is decided by whether the outer function is `async` or not. There are no synchronous or asynchronous commands in Nightwatch and all the commands return a promise if the outer function is `async` and no command returns a promise if the outer function is not `async`. For example:

```js
it('synchronous test', function() {
  const result1 = browser.getText('#main ul li a.first');
  const result2 = browser.navigateTo('https://nightwatchjs.org');

  console.log(result1 instanceof Promise);  // false
  console.log(result2 instanceof Promise);  // false
});

it('asynchronous test', async function() {
  const result1 = browser.getText('#main ul li a.first');
  const result2 = browser.navigateTo('https://nightwatchjs.org');

  console.log(result1 instanceof Promise);  // true
  console.log(result2 instanceof Promise);  // true
});
```
